compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorDistribute.cpp

```diff
@@ -0,0 +1,333 @@
+    return failure();
+  }
+  int64_t residualElements = problemSize;
+  for (int i = sizes.size() - 1, e = 1; i >= e; --i) {
+    if (sizes[i] <= 0) {
+      return failure();
+    }
+    if (residualElements % sizes[i] != 0) {
+      return failure();
+    }
+    residualElements /= sizes[i];
+  }
+  sizes[0] = residualElements;
+  return PerDimLayoutAttr::get(context, dimTypes, sizes);
+}
+
+// Vector layout option setter aimed at contractions. Currently this only sets
+// anchors for two types of operation; vector.contract and vector.transfer_read
+// from non-shared memory. The assumption in this case is that all IR input to
+// this pass has a leaf rooted on a transfer_read or includes a contraction in
+// the program slice, meaning all operations should receive layouts. Layout
+// setting for other problems like reductions is TODO.
+class ContractionVectorLayoutOptions : public VectorLayoutOptions {
+public:
+  ContractionVectorLayoutOptions(Operation *root, ArrayAttr types,
+                                 ArrayRef<int64_t> workgroupSize, Value laneId)
+      : VectorLayoutOptions(root), mmaTypes(types),
+        workgroupSize(workgroupSize), patterns(root->getContext()) {
+    populateGPUDistributionPatterns(patterns);
+    populateGPUDistributionLayoutAttrPatterns(laneId, patterns);
+  }
+
+  void setAnchorOps(VectorLayoutAnalysis &analysis) override {
+    MLIRContext *context = root->getContext();
+    root->walk([&](Operation *op) {
+      llvm::TypeSwitch<Operation *>(op)
+          .Case([&](vector::ContractionOp contract) {
+            setContractionAnchor(context, analysis, contract);
+          })
+          .Case([&](vector::TransferReadOp transfer) {
+            setTransferReadAnchor(context, analysis, transfer);
+          });
+    });
+  }
+
+  RewritePatternSet &getPatterns() { return patterns; }
+
+private:
+  // Sets an anchoring layout for the given contraction op. Looks for a
+  // supported mma type from the cached list of mma types and populates the
+  // necessary distribution pattern for those contractions.
+  void setContractionAnchor(MLIRContext *context,
+                            VectorLayoutAnalysis &analysis,
+                            vector::ContractionOp contract) {
+    FailureOr<IREE::GPU::MmaAttr> maybeMmaType =
+        getCompatibleMmaAttr(mmaTypes, contract);
+    // TODO: Add SIMT fallback.
+    assert(succeeded(maybeMmaType) && "incompatible contraction op");
+
+    auto mmaType = *maybeMmaType;
+    auto maybeLayouts = mmaType.getContractionLayout(contract);
+    assert(succeeded(maybeMmaType) && "mma layout type must not be opaque");
+
+    auto [aLayout, bLayout, cLayout] = *maybeLayouts;
+    analysis.setAnchor(contract.getLhs(), aLayout);
+    analysis.setAnchor(contract.getRhs(), bLayout);
+    analysis.setAnchor(contract.getAcc(), cLayout);
+    analysis.setAnchor(contract.getResult(), cLayout);
+
+    if (isa<IREE::GPU::MFMAAttr>(mmaType)) {
+      if (!populatedMfma) {
```

Comment:
This should be gated on the executable being for rocm?

Reply:
I'm using MFMAAttr to mean "use `amdgpu.mfma`" so we can just set it in ROCMExecutableFeatures rather than having to verify what the executable is being compiled for from within this pass. It would be an error if a SPIR-V target said they support an MFMA attribute, but the error would happen during conversion to SPIR-V because the target environment was wrong. I did it this way because I didn't want to reach for target chip information after kernel config. If we want, we could add additional verification to the hal.executable target attribute; no need to wait until we try to codegen + distribution to emit a failure for an invalid target.